File: packages/sdk-py/src/maestro_sdk/task.py
from typing import Generic, Protocol, TypeVar

from .types import Json, RunContext

TIn = TypeVar("TIn")
TOut = TypeVar("TOut")


class Task(Protocol, Generic[TIn, TOut]):
    def init(self, ctx: RunContext) -> None: ...  # optional
    def validate(self, input: Json) -> None: ...  # optional
    def execute(self, ctx: RunContext, input: Json) -> Json: ...


def define_task(task: Task[TIn, TOut]) -> Task[TIn, TOut]:
    return task

File: packages/sdk-ts/src/task.ts
import type { RunContext, TaskInput, TaskOutput } from './types.js';

export interface Task<TIn = unknown, TOut = unknown> {
  init?: (ctx: RunContext) => Promise<void> | void;
  validate?: (input: TaskInput<TIn>) => Promise<void> | void;
  execute: (
    ctx: RunContext,
    input: TaskInput<TIn>,
  ) => Promise<TaskOutput<TOut>>;
}

export function defineTask<TIn = unknown, TOut = unknown>(
  task: Task<TIn, TOut>,
): Task<TIn, TOut> {
  return task;
}
